src/RecursiveIteration.php

```diff
@@ -0,0 +1,73 @@
+<?php
+
+namespace Dhii\Iterator;
+
+use Dhii\Util\String\StringableInterface as Stringable;
+use Traversable;
+
+/**
+ * Represents an iteration in a recursive iterator.
+ *
+ * @since [*next-version*]
+ */
+class RecursiveIteration extends AbstractBaseIteration implements RecursiveIterationInterface
+{
+    /**
+     * An array of path segments.
+     *
+     * @since [*next-version*]
+     *
+     * @var array
+     */
+    protected $path;
+
+    /**
+     * Constructor.
+     *
+     * @since [*next-version*]
+     *
+     * @param string|int|null                   $key   The iteration key.
+     * @param mixed                             $value The iteration value.
+     * @param string[]|Stringable[]|Traversable $path  A list of path segments.
+     */
+    public function __construct($key, $value, $path = [])
+    {
+        $this->_setKey($key)
+             ->_setValue($value)
+             ->_setPathSegments($path);
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @since [*next-version*]
+     */
+    public function getPathSegments()
+    {
+        return $this->path;
+    }
+
+    /**
+     * Sets the path segments.
+     *
+     * @since [*next-version*]
+     *
+     * @param string[]|Stringable[]|Traversable $path A list of path segments.
+     *
+     * @return $this
+     */
+    protected function _setPathSegments($path)
+    {
+        $this->path = $path;
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @since [*next-version*]
+     */
+    public function getDepth()
+    {
+        return count($this->getPathSegments()) - 1;
```

Comment:
@mecha, I really find it hard to follow the logic you described, because to me the logic seems sound, but the conclusion seems wrong.

In the example I gave, where the path is `C -> D -> E`, you are in node `A`, and therefore, you have the following transitions:

1. `A -> C`
2. `C -> D`
3. `D -> E`

That's a total of 3 transitions. For a path of 3 segments. 3 = 3. Number of transitions equals the number of segments.